src/App.tsx

```diff
@@ -1,18 +1,59 @@
-import React from 'react'
-import './index.css'
-import { FaCheckCircle } from 'react-icons/fa'
-import TodoList from './components/TodoList'
+import React from 'react';
+import { useState } from 'react';
+import './index.css';
+import { FaCheckCircle, FaLevelUpAlt } from 'react-icons/fa';
+import TodoList from './components/TodoList';
+import { TodoProps } from './interface';
+import TodoAdd from './components/TodoAdd';
+import { v4 as uuidv4 } from 'uuid';
+
+const App: React.FC = () => {
+  const [todos, setTodo] = useState<TodoProps[]>([]);
+
+  // this is redundant
+  // how can I write this to not mutate the value in place
+  const handleTaskComplete = (id: string) => {
+    const newTodoState: TodoProps[] = [...todos];
```

Comment:
Though confusing, I think this syntax functionally does a copy - so you're not technically mutating the value in place

Reply:
Oh you mean not setting the `isComplete` property, but instead creating a new task - also a little confusing, but you could do something like:

```ts
const index = newTodoState.findIndex((todo) => todo.id === id)
const todo = newTodoState[index]
newTodoState[index] = Object.assign({ isComplete: !todo.isComplete}, todo) // creates a copy
```

ooor if you wanna really be strict about immutability you can do:

```ts
setTodo(
    ...newTodoState.slice(0, index),
    Object.assign({ isComplete: !todo.isComplete }, todo),
   ...newTodoState.slice(index + 1),
);
```

Or if you want to use the more clever language syntax you can try the approach [suggested here](https://stackoverflow.com/a/38061256):
```ts
setTodo(
   Object.assign([], newTodoState, {
      index: Object.assign({ isComplete: !todo.isComplete }, todo),
   }),
)
```
For the purest example of immutability 😭

Reply:
Ah I see, you manually set the `isComplete` property. I think that's probably fine - in practice if you wanted to change TodoProps to a class you could add factory functions to it.